go/teams/teams.go

```diff
@@ -1474,3 +1474,164 @@ func (t *Team) associateTLFID(ctx context.Context, tlfID keybase1.TLFID) (err er
+// adding members:
+// - User can be PUK-less. AddMembersBestEffort will create a
+//   keybase-type invite in this case.
+// - This user might have reset and previous UV is in the team.
+//   Then, a delete will be issued for that UV before adding new
+//   membership (or adding a new invite).
+// - If user has a PUK, it will add them normally, unless they
+//   already are in the team with a higher role.
+//
+// This function can be called multiple times with the same UV list
+// with no side effects. Caller is responsible for handling cases when
+// this function races other clients and hits bad sigchain sequence
+// numbers. `RetryOnSigOldSeqnoError` can be used for this. Even if
+// two clients race with adding overlapping UV sets, they should
+// eventually reconcile with all expected members in.
+//
+// This function only returns an error in the following conditions:
+// - Team fails to load.
+// - `role` argument is invalid.
+// - Either ChangeMembership or Invite signature fails to post
+//   (e.g. because of race with other client - see above.)
+//
+// If individual user cannot be processed, errors are only logged,
+// and the process continues.
+func AddMembersBestEffort(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, role keybase1.TeamRole, uvs []keybase1.UserVersion, forceRepoll bool) (err error) {
+	defer g.CTrace(ctx, fmt.Sprintf("AddMembersBestEffort(TeamID:%s,role:%v)", teamID, role), func() error { return err })()
+
+	team, err := Load(ctx, g, keybase1.LoadTeamArg{
+		ID:          teamID,
+		Public:      teamID.IsPublic(),
+		ForceRepoll: forceRepoll,
+		NeedAdmin:   true,
+	})
+	if err != nil {
+		return err
+	}
+
+	var req keybase1.TeamChangeReq
+	var keybaseInvites []keybase1.UserVersion
+	var needPostMembership bool
+
+	for _, uv := range uvs {
+		var becameAnInvite bool
+		loadedUV, err := loadUserVersionByUID(ctx, g, uv.Uid)
+		if err != nil {
+			if err == errInviteRequired {
+				keybaseInvites = append(keybaseInvites, uv)
+				becameAnInvite = true
+				g.Log.CDebugf(ctx, "Invite required for %+v", uv)
+			} else {
+				g.Log.CDebugf(ctx, "got unexpected error while trying to load user: %s: %v; skipping", uv, err)
+				continue
+			}
+		}
+
+		if loadedUV.EldestSeqno != uv.EldestSeqno {
+			// TODO: We can skip this one or correct the EldestSeqno
+			// in the request.
+			g.Log.CDebugf(ctx, "Trying to add UID %s with EldestSeqno %d, but servers says EldestSeqno is %d", uv.Uid, uv.EldestSeqno, loadedUV.EldestSeqno)
+			continue
+		}
+
+		currentRole, err := team.MemberRole(ctx, uv)
+		if err != nil {
+			g.Log.CDebugf(ctx, "team.MemberRole(%+v) failed with %+v", uv, err)
+			continue
+		}
+
+		if currentRole.IsOrAbove(role) {
+			g.Log.CDebugf(ctx, "User %s already has role %v, skipping", uv.Uid, currentRole)
+			continue
+		}
+
+		existingUV, err := team.UserVersionByUID(ctx, uv.Uid)
+		if err == nil {
+			if existingUV.EldestSeqno > uv.EldestSeqno {
```

Comment:
@mlsteele  Do you think adding `uv.EldestSeqno != 0 &&` to this condition solves one of the edge cases?

Reply:
or I guess `!becameAnInvite &&` - we might want to clear all person's cryptomembers if we discover person that we are adding is pukless dude. I don't know which one of these is sound :|

Reply:
Your suggestion of this sounds like it will work
 ```
if !becameAnInvite && existingUV.EldestSeqno > uv.EldestSeqno
```